tests/functional/test_snap.py

```diff
@@ -1,5 +1,153 @@
+import json
 import subprocess
+import urllib.request
+from contextlib import contextmanager
 
+# import pytest
+from tenacity import retry, stop_after_delay, wait_fixed
 
-def test_snap_installed():
-    subprocess.check_call(["snap", "list", "smartctl-exporter"])
+ENDPOINT = "http://localhost:9633/metrics"
+SYSTEMD_SERVICE = "snap.smartctl-exporter.smartctl-exporter"
+SNAP_NAME = "smartctl-exporter"
+
+
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _check_service_failed() -> None:
+    """Check if a systemd service is in a failed state."""
+    assert 0 == subprocess.call(
+        f"sudo systemctl is-failed --quiet {SYSTEMD_SERVICE}".split()
+    ), f"{SYSTEMD_SERVICE} is running"
+
+
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _check_service_active() -> None:
+    """Check if a systemd service is active."""
+    assert 0 == subprocess.call(
+        f"sudo systemctl is-active --quiet {SYSTEMD_SERVICE}".split()
+    ), f"{SYSTEMD_SERVICE} is not running"
+
+
+@retry(wait=wait_fixed(5), stop=stop_after_delay(30))
+def _check_endpoint(endpoint: str) -> None:
+    """Check if an endpoint is reachable."""
+    response = urllib.request.urlopen(endpoint)  # will raise if not reachable
+    status_code = response.getcode()
+    assert status_code == 200, f"Endpoint {endpoint} returned status code {status_code}"
+
+
+@retry(wait=wait_fixed(5), stop=stop_after_delay(30))
+def _get_endpoint_data(endpoint: str) -> str:
+    """Get the data from an endpoint."""
+    response = urllib.request.urlopen(endpoint)
+    return response.read().decode("utf-8")
+
+
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _check_bind_config(bind: str) -> None:
+    """Check if a service is listening on a specific bind."""
+    assert 0 == subprocess.call(
+        f"nc -z localhost {bind.lstrip(':')}".split()
```

Comment:
suggestion: while within the context of these tests this is adequate, it's not technically telling you if the snap is the one listening on that port. You could consider using lsof for a more precise check. Something like `lsof -sTCP:LISTEN -i tcp:<bind port>` should give you the process that is listening on that specific port. Further output filtering switches may be available.